### PR TITLE
Substantial code refactoring.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Splitting classes into separate files.
Remove unnecessary or redundant code.
General code clean-up.
Creating a 'Settings' class to hold all the command line argument options that are passed into dotnet.

### CODE REVIEW

File: demos/dotnet/dotnet/CscAction.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+using System.IO;
+using System.Text;
+
+namespace dotnet
+{
+    internal static class CscAction
+    {
+        public static bool Execute(ProjectProperties properties, Log log)
+        {
+            Console.WriteLine("compiling");
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/CscAction.cs]: Why is this not the log?
It's not really compiling, right? "Invoking Compiler"?

File: demos/dotnet/dotnet/CscAction.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            var processSettings = new ProcessStartInfo
+            {
+                FileName = properties.CscPath,
+                Arguments = properties.FormatCscArguments()
+            };
+
+            log.WriteLine("Executing {0}", processSettings.FileName);
+            log.WriteLine("Csc Arguments: {0}", processSettings.Arguments);
+
+            processSettings.CreateNoWindow = true;
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/CscAction.cs]: Why are these properties not set above when you create processSettings? Seems like the initializer can own this concern all-up...

File: demos/dotnet/dotnet/CscAction.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            processSettings.UseShellExecute = false;
+
+            Process cscProcess;
+            try
+            {
+                cscProcess = Process.Start(processSettings);
+            }
+            catch (Win32Exception)
+            {
+                Console.WriteLine("ERROR: csc.exe needs to be on the path.");
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/CscAction.cs]: Log.WriteLine?

Comment (piotrpMSFT) [inline demos/dotnet/dotnet/CscAction.cs]: that message seems to conflict with properties.CscPath. Why is there a path on properties but it also needs to be on "the path" which I assume means the PATH env var?

File: demos/dotnet/dotnet/CscAction.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            if (cscProcess == null) return false;
+            var output = cscProcess.StandardOutput.ReadToEnd();
+            log.WriteLine(output);
+
+            cscProcess.WaitForExit();
+
+            return !output.Contains("error CS");
+        }
+
+        private static string FormatReferenceOption(this ProjectProperties project)
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/CscAction.cs]: This doesn't seem like the right place for these ProjectProperties formatters. ProjectPropertyCscExtensions or something...?

File: demos/dotnet/dotnet/CscAction.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                {
+                    first = false;
+                }
+                builder.Append(Path.Combine(project.PackagesDirectory, reference));
+            }
+            builder.Append(" ");
+            return builder.ToString();
+        }
+
+        private static string FormatSourcesOption(this ProjectProperties project)
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/CscAction.cs]: same as above...

File: demos/dotnet/dotnet/CscAction.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            foreach (var source in project.Sources)
+            {
+                builder.Append(" ");
+                builder.Append(source);
+            }
+            return builder.ToString();
+        }
+
+        private static string FormatCscOptions(this ProjectProperties project)
+        {
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/CscAction.cs]: And here...

File: demos/dotnet/dotnet/CscAction.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+            builder.Append(" ");
+            builder.Append("/out:");
+            builder.Append(project.OutputAssemblyPath);
+            builder.Append(" ");
+            return builder.ToString();
+        }
+
+        private static string FormatCscArguments(this ProjectProperties project)
+        {
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/CscAction.cs]: And here...

File: demos/dotnet/dotnet/Helpers.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+
+namespace dotnet
+{
+    internal static class Helpers
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/Helpers.cs]: These are System.IO.File extensions? FileSystemHelpers? Helpers seems like a dumping-ground sort of name...

File: demos/dotnet/dotnet/Log.cs [hunk 9]
```diff
@@ -4,39 +4,42 @@
 using System;
 using System.Collections.Generic;
 
-class Log
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/Log.cs]: Can we put this behind an interface? I don't imagine this is the only logger folks might want...

File: demos/dotnet/dotnet/CscAction.cs (hunk 1 above)
Comment (ahsonkhan) [inline demos/dotnet/dotnet/CscAction.cs]: Already fixed for next check-in.
I will change the string.

File: demos/dotnet/dotnet/CscAction.cs (hunk 2 above)
Comment (ahsonkhan) [inline demos/dotnet/dotnet/CscAction.cs]: Agreed.

File: demos/dotnet/dotnet/NugetAction.cs [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http;
+using System.Text;
+
+namespace dotnet
+{
+    internal static class NugetAction
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: These class-level Action concepts seem like they'd be better named Agents or similar. They are not actions themselves, instead being containers/groupings of related actions. This is especially visible in the method of this Type called "DownloadNugetAction" leading to "NugetAction.DownloadNugetAction()". 

I think I'd rather see "NugetAgent.DownloadNuget()". Thoughts?

File: demos/dotnet/dotnet/NugetAction.cs [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    {
+        public static bool GetNugetAndRestore(ProjectProperties properties, Log log)
+        {
+            var nugetFile = Path.Combine(properties.ToolsDirectory, "nuget.exe");
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: This feels like it should follow your Csc extension method pattern. Something like properties.GetNugetExePath()...

File: demos/dotnet/dotnet/NugetAction.cs [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+            var success = true;
+            if (!File.Exists(nugetFile))
+            {
+                success = DownloadNugetAction(properties, nugetFile);
+            }
+
+            var projectJsonFile = Path.Combine(properties.ProjectDirectory, "project.json");
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: properties.GetProjectJsonPath()?

File: demos/dotnet/dotnet/NugetAction.cs [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+using System.IO;
+using System.Net.Http;
+using System.Text;
+
+namespace dotnet
+{
+    internal static class NugetAction
+    {
+        public static bool GetNugetAndRestore(ProjectProperties properties, Log log)
+        {
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: This method seems like it's really this:

    {
        EnsureNugetExe();
        RestoreProjectDependencies(properties);
    }

Is that the right way to factor it?

File: demos/dotnet/dotnet/NugetAction.cs [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+            var success = true;
+            if (!File.Exists(nugetFile))
+            {
+                success = DownloadNugetAction(properties, nugetFile);
+            }
+
+            var projectJsonFile = Path.Combine(properties.ProjectDirectory, "project.json");
+            if (!File.Exists(projectJsonFile))
+            {
+                projectJsonFile = Path.Combine(properties.ToolsDirectory, "project.json");
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: Why would project.json be in the ToolsDirectory?

Seems you want a GetProjectJsonOrDefault()?

File: demos/dotnet/dotnet/CscAction.cs (hunk 3 above)
Comment (ahsonkhan) [inline demos/dotnet/dotnet/CscAction.cs]: If the csc.exe is not found in the properties.CscPath, the CscPath property is default set to 'csc.exe'. If csc.exe is not found anywhere up the path (from the current directory), this error is reported. Where we get csc.exe needs to be fixed as right now it basically just looks for it anywhere within .NET framework and hopes one exists. Once that is fixed, I think, that is when we can address proper error handling. Thoughts?

File: demos/dotnet/dotnet/NugetAction.cs (hunk 10 above)
Comment (ahsonkhan) [inline demos/dotnet/dotnet/NugetAction.cs]: Agreed, that makes sense to me.

File: demos/dotnet/dotnet/NugetAction.cs (hunk 14 above)
Comment (ahsonkhan) [inline demos/dotnet/dotnet/NugetAction.cs]: The project.json in the ToolsDirectory is the auto-generated one.

File: demos/dotnet/dotnet/CscAction.cs (hunk 3 above)
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/CscAction.cs]: Sounds complicated. Can we encapsulate the CSC path finding stuff somewhere?

File: demos/dotnet/dotnet/NugetAction.cs (hunk 14 above)
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: Is that auto-generated one constant? Or does it change?

File: demos/dotnet/dotnet/NugetAction.cs [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+            {
+                Console.WriteLine("Failed to get nuget or restore packages.");
+                return false;
+            }
+            return true;
+        }
+
+        public static bool DownloadNugetAction(ProjectProperties properties, string nugetFile)
+        {
+            CreateDefaultProjectJson(properties);
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: Why are these two methods here? I'd not expect that from "Download Nuget"

File: demos/dotnet/dotnet/NugetAction.cs [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+            CreateNugetConfig(properties);
+
+            var client = new HttpClient();
+            var requestUri = new Uri(@"http://dist.nuget.org/win-x86-commandline/v3.1.0-beta/nuget.exe",
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: Probably later, but this should be configurable.

File: demos/dotnet/dotnet/NugetAction.cs [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+                UriKind.Absolute);
+
+            var sourceStreamTask = client.GetStreamAsync(requestUri);
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: Is WebClient not available to us? DownloadFile seems cleaner...

File: demos/dotnet/dotnet/NugetAction.cs [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+            try
+            {
+                sourceStreamTask.Wait();
+            }
+            catch (AggregateException exception)
+            {
+                foreach (var ex in exception.InnerExceptions)
+                {
+                    Console.WriteLine(ex.Message);
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: log.Writeline?

File: demos/dotnet/dotnet/NugetAction.cs [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+                UriKind.Absolute);
+
+            var sourceStreamTask = client.GetStreamAsync(requestUri);
+            try
+            {
+                sourceStreamTask.Wait();
+            }
+            catch (AggregateException exception)
+            {
+                foreach (var ex in exception.InnerExceptions)
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: Should we dump all exceptions? Or just say that download has failed?

File: demos/dotnet/dotnet/NugetAction.cs [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+                    var read = sourceStream.Read(buffer, 0, buffer.Length);
+                    if (read < 1) break;
+                    destinationStream.Write(buffer, 0, read);
+                }
+            }
+
+            return true;
+        }
+
+        private static void CreateDefaultProjectJson(ProjectProperties properties)
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: Why is the NugetAgent creating the Project.JSON?

File: demos/dotnet/dotnet/NugetAction.cs [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+        {
+            var fileName = Path.Combine(properties.ToolsDirectory, "project.json");
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: properties.GetProjectJsonPath()

File: demos/dotnet/dotnet/NugetAction.cs [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+                        file.WriteLine();
+                    }
+                }
+                file.WriteLine(@"    },");
+                file.WriteLine(@"    ""frameworks"": {");
+                file.WriteLine(@"        ""dnxcore50"": { }");
+                file.WriteLine(@"    }");
+                file.WriteLine(@"}");
+
+                //"runtimes": {
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: delete these ;)

File: demos/dotnet/dotnet/NugetAction.cs [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+                //"win7-x86": { },
+                //"win7-x64": { }
+                //},
+            }
+        }
+
+        private static void CreateNugetConfig(ProjectProperties properties)
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: If this were nuget.exe, is it responsible for Create Nuget Config? Or is that someone else's job?

File: demos/dotnet/dotnet/NugetAction.cs [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+                file.WriteLine(@"    </packageSources>");
+                file.WriteLine(@"</configuration>");
+            }
+        }
+
+        public static bool RestorePackagesAction(ProjectProperties properties, Log log, string nugetFile, string jsonFile)
+        {
+            Console.WriteLine("restoring packages");
+
+            if (!File.Exists(nugetFile))
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: We already have a method for ensuring NuGet exists, right?

File: demos/dotnet/dotnet/NugetAction.cs (hunk 13 above)
Comment (ahsonkhan) [inline demos/dotnet/dotnet/NugetAction.cs]: Sure, that is the intent.

File: demos/dotnet/dotnet/NugetAction.cs [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+                file.WriteLine(@"        <add key = ""nuget.org"" value = ""https://www.nuget.org/api/v2/""/>");
+
+                file.WriteLine(@"    </packageSources>");
+                file.WriteLine(@"</configuration>");
+            }
+        }
+
+        public static bool RestorePackagesAction(ProjectProperties properties, Log log, string nugetFile, string jsonFile)
+        {
+            Console.WriteLine("restoring packages");
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: log.WriteLine()?

File: demos/dotnet/dotnet/NugetAction.cs [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+
+            if (!File.Exists(nugetFile))
+            {
+                Console.WriteLine("Could not find file {0}.", nugetFile);
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: log.WriteLine()?

File: demos/dotnet/dotnet/NugetAction.cs (hunk 11 above)
Comment (ahsonkhan) [inline demos/dotnet/dotnet/NugetAction.cs]: I don't understand what you mean by that.

File: demos/dotnet/dotnet/NugetAction.cs [hunk 27]
```diff
@@ -0,0 +1,188 @@
...
+                return false;
+            }
+
+            if (!File.Exists(jsonFile))
+            {
+                Console.WriteLine("Could not find file {0}.", jsonFile);
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: log.WriteLine()?

File: demos/dotnet/dotnet/NugetAction.cs [hunk 28]
```diff
@@ -0,0 +1,188 @@
...
+                return false;
+            }
+
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/NugetAction.cs]: For all of the above, it seems we'd want to throw an Argument Exception, no?

File: demos/dotnet/dotnet/NugetAction.cs (hunk 12 above)
Comment (ahsonkhan) [inline demos/dotnet/dotnet/NugetAction.cs]: To implement it the way you are suggesting would require a bit more work. Can we add that as a separate issue?

File: demos/dotnet/dotnet/OtherActions.cs [hunk 29]
```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Text;
+
+namespace dotnet
+{
+    internal static class OtherActions
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/OtherActions.cs]: This seems like another dumping-ground class. This class represents the ProjectAgent and the BuildAgent... no?

File: demos/dotnet/dotnet/ProjectProperties.cs [hunk 30]
```diff
@@ -17,15 +18,9 @@ class ProjectProperties
         public readonly List<string> CscOptions = new List<string>();
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectProperties.cs]: Looks like this class is specialized to CSC... why not put the CSC extension methods in here, along with beefing up the name?

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 31]
```diff
@@ -0,0 +1,335 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Xml;
+
+namespace dotnet
+{
+    internal static class ProjectPropertiesHelpers
+    {
+        public static ProjectProperties InitializeProperties(Settings settings, Log log)
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: Why isn't this a constructor?

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 32]
```diff
@@ -0,0 +1,335 @@
...
+        {
+            // General Properites
+            var properties = new ProjectProperties();
+            var currentDirectory = Directory.GetCurrentDirectory();
+
+            var buildDll = settings.Target == "library";
+
+            properties.ProjectDirectory = Path.Combine(currentDirectory);
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: initializer?

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 33]
```diff
@@ -0,0 +1,335 @@
...
+            properties.PackagesDirectory = Path.Combine(properties.ProjectDirectory, "packages");
+            properties.OutputDirectory = Path.Combine(properties.ProjectDirectory, "bin");
+            properties.ToolsDirectory = Path.Combine(properties.ProjectDirectory, "tools");
+            properties.AssemblyName = Path.GetFileName(properties.ProjectDirectory);
+            properties.OutputType = buildDll ? ".dll" : ".exe";
+            FindCompiler(properties);
+
+            AddToListWithoutDuplicates(properties.Sources, ParseProjectFile(properties, settings.ProjectFile, "Compile"));
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: Why not use a Set?

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 34]
```diff
@@ -0,0 +1,335 @@
...
+            var currentDirectory = Directory.GetCurrentDirectory();
+
+            var buildDll = settings.Target == "library";
+
+            properties.ProjectDirectory = Path.Combine(currentDirectory);
+            properties.PackagesDirectory = Path.Combine(properties.ProjectDirectory, "packages");
+            properties.OutputDirectory = Path.Combine(properties.ProjectDirectory, "bin");
+            properties.ToolsDirectory = Path.Combine(properties.ProjectDirectory, "tools");
+            properties.AssemblyName = Path.GetFileName(properties.ProjectDirectory);
+            properties.OutputType = buildDll ? ".dll" : ".exe";
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: Is this guaranteed to be one of the two? do we not want an enum?

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 35]
```diff
@@ -0,0 +1,335 @@
...
+            FindCompiler(properties);
+
+            AddToListWithoutDuplicates(properties.Sources, ParseProjectFile(properties, settings.ProjectFile, "Compile"));
+            
+            foreach (var file in settings.SourceFiles.Where(f => !f.StartsWith(properties.PackagesDirectory) && !f.StartsWith(properties.OutputDirectory) &&
+                                                              !f.StartsWith(properties.ToolsDirectory)))
+            {
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: this condition is pretty gnarly... what is it meant to represent?

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 36]
```diff
@@ -0,0 +1,335 @@
...
+                AddToListWithoutDuplicates(properties.Sources, file);
+            }
+
+            if (!string.IsNullOrWhiteSpace(settings.Recurse) && settings.Recurse != "*.cs")
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: This deserves a name as well... not sure what's going on!

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 37]
```diff
@@ -0,0 +1,335 @@
...
+            {
+                AddToListWithoutDuplicates(properties.Sources,
+                    Directory.GetFiles(properties.ProjectDirectory, settings.Recurse, SearchOption.AllDirectories));
+            }
+
+            if (properties.Sources.Count == 1)
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: else...?

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 38]
```diff
@@ -0,0 +1,335 @@
...
+            {
+                properties.AssemblyName = Path.GetFileNameWithoutExtension(properties.Sources[0]);
+            }
+
+            var platformOptionSpecicifcation = GetPlatformOption(settings.Platform);
+
+            // The anycpu32bitpreferred setting is valid only for executable (.EXE) files
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: here down, I think most places where you have a comment we should break out that block of code into a method.

Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: We should also build a list of strings, then add the preamble to that list of strings. It's more work to understand the code like this and there is more opportunity for error.

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 39]
```diff
@@ -0,0 +1,335 @@
...
+            AddToListWithoutDuplicates(properties.Packages, ParseProjectFile(properties, settings.ProjectFile, "Package"));
+
+            LogProperties(log, properties, "Adjusted Properties Log:", buildDll);
+
+            return properties;
+        }
+
+        public static string GetPlatformOption(string optionSpecification)
+        {
+            if (optionSpecification == "anycpu32bitpreferred" || optionSpecification == "x86")
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: switch(optionSpecification):?

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 40]
```diff
@@ -0,0 +1,335 @@
...
+            {
+                return "x86";
+            }
+            return "x64";
+        }
+
+        public static string GetRuntimeCoreClr(string platform)
+        {
+            // platform - x86, x64, arm
+            return "\"Microsoft.NETCore.Runtime.CoreCLR-" + platform.ToLower() + "\": \"1.0.0\"";
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: from here down, consider string.Format()

Better yet, consider C# 6 string interpolation: https://msdn.microsoft.com/en-us/magazine/dn879355.aspx

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 41]
```diff
@@ -0,0 +1,335 @@
...
+            log.WriteLine("output path          {0}", project.OutputAssemblyPath);
+            log.WriteList(project.Sources, "SOURCES");
+            log.WriteList(project.Packages, "PACKAGES");
+            log.WriteList(project.References, "REFERENCES");
+            log.WriteList(project.Dependencies, "DEPENDENCIES");
+            log.WriteList(project.CscOptions, "CSCOPTIONS");
+            log.WriteLine("-------------------------------------------------");
+        }
+
+        private static void Adjust(ProjectProperties properties, string adjustmentFilePath, ICollection<string> list)
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: name is unclear...

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 42]
```diff
@@ -0,0 +1,335 @@
...
+                    list.Add(Path.Combine(properties.PackagesDirectory, adjustment));
+                }
+            }
+        }
+
+        private static List<string> ParseProjectFile(ProjectProperties properties, string projectFile,
+            string elementName)
+        {
+            var attributes = GetAttributes(elementName);
+            if (!File.Exists(projectFile) || attributes == null)
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: initialize attributeValues before this block. Then just return it if the condition is met.

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 43]
```diff
@@ -0,0 +1,335 @@
...
+                    list.Remove(Path.Combine(properties.PackagesDirectory, adjustment));
+                }
+                else
+                {
+                    list.Add(Path.Combine(properties.PackagesDirectory, adjustment));
+                }
+            }
+        }
+
+        private static List<string> ParseProjectFile(ProjectProperties properties, string projectFile,
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: Think about the intent of this method. It seems to be building a list of Xml attributes for nodes with a particular name... that's not what I thinjk of when I read "ParseProjectFile"...

Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: Ok, it does some other stuff with the values depending on the node type, but you understand the concern I'm sure.

File: demos/dotnet/dotnet/ProjectPropertiesHelpers.cs [hunk 44]
```diff
@@ -0,0 +1,335 @@
...
+                default:
+                    values = new List<string>();
+                    break;
+            }
+            return values;
+        }
+
+        private static List<string> GetAttributes(string elementName)
+        {
+            var compileAttributes = new List<string> {"Include"};
```
Comment (piotrpMSFT) [inline demos/dotnet/dotnet/ProjectPropertiesHelpers.cs]: Can these be const?

**Comment by ahsonkhan** [comment]
I will create a work item to address some of the comments made here as they are important but would take time. I will continue and merge the changes however since additional work depends on some of these changes.

**Pull request merged**